### PR TITLE
Selector reimplemented in C using epoll/select in a spinlock-free way

### PR DESCRIPTION
**Pull request by skateman**
None of the previous implementations of `surro-gate` were completely spinlock-free. Waiting for a pair of sockets to be ready to transmit is not what `nio4r` is designed for. I went a few layers deeper and ended up at epoll's EPOLLONESHOT event (EV_ONESHOT in the kqueue world). Thanks to this feature a registered socket wakes up up the waiting function only once until it is not reset.

The proxy stores two `SurroGate::Pair` objects for each socket pair, each of them represents a transmission direction with readiness for read and write between the two sockets. Meanwhile there's a single selector (epoll or kqueue) with all the sockets registered for both read and write events and pointers to the related `SurroGate::Pair` objects. If the selector returns with an event, the related pairing objects' readiness will be updated accordingly and rearmed in the opposite direction if necessary.

The transmission handling should be done by passing a block to the `each_ready` method. After the block is called, the readiness is reset and the sockets are rearmed accordingly. Pairs with at least one closed socket are also passed to this block, so closing them and cleaning up can be done using exceptions (see in the README.md below).

The fallback-select implementation is awful, slow and ineffective, but as this library is primarily intended to use on Linux, it is suitable for now. In the future it might be refactored and/or optimized.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -33,14 +32,29 @@ require 'surro-gate'
 proxy = SurroGate.new
 
 # Create two TCP socket connections
```
Comment (kbrock) [inline README.md]: 2 sockets, one for each direction

File: README.md [hunk 2]
```diff
@@ -33,14 +32,29 @@ require 'surro-gate'
...
-proxy.push(left, right)
-
-# Wait for the communication to finish
-proxy.wait
+proxy.push(sock_1, sock_2)
+proxy.push(sock_3, sock_4)
+
+loop do
+  # Call the selector with a timeout value in milliseconds
+  proxy.select(1000)
```
Comment (kbrock) [inline README.md]: select with timeout after 1 second

File: README.md [hunk 3]
```diff
@@ -33,14 +32,29 @@ require 'surro-gate'
...
+
+  # Do some hard work
+  proxy.each_ready do |left, right|
+    begin
+      right.write_nonblock(left.read_nonblock(4096))
```
Comment (kbrock) [inline README.md]: doesn't this return the count of characters written?

File: ext/surro-gate/proxy.c [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+  selector_register(selector, SOCK_PTR(right), index_rtl, index_ltr);
+
+  return Qtrue;
+}
+
+static VALUE SurroGate_Proxy_pop(VALUE self, VALUE sockets) {
+  int i, *selector;
+
+  VALUE pairing = rb_iv_get(self, "@pairing");
+  rb_block_call(pairing, rb_intern("delete_if"), 0, NULL, pairing_compare, sockets); // @pairing.delete_if(&pairing_compare)
```
Comment (kbrock) [inline ext/surro-gate/proxy.c]: Little confused why there may be multiple sockets here.
maybe confused what exactly are sockets - are there 2  - one for each direction?

File: README.md (hunk 1 above)
Comment (skateman) [reply README.md]: no, 2 sockets to proxy between them

File: README.md (hunk 3 above)
Comment (skateman) [reply README.md]: it does

File: ext/surro-gate/proxy.c (hunk 4 above)
Comment (skateman) [reply ext/surro-gate/proxy.c]: So you have two sockets, let's say `a` and `b`. If you push these into the proxy, you will have 2 pairing objects `a->b` and `b->a`.

File: README.md (hunk 3 above)
Comment (kbrock) [reply README.md]: lol. And aren't you curious if all characters went through?

File: README.md (hunk 1 above)
Comment (kbrock) [reply README.md]: aah cool. Could you change comment for this?

**Comment by skateman** [comment]
@kbrock I got the ruby implementation for mac, do you have some time to look at it? :wink: :tada:

**Comment by kbrock** [comment]
I tested with the following configuration. thanks

```ruby
# terminal 1: nc -clp 1111
# terminal 2: nc -clp 1112
# terminal 3: ruby -I . this_script.rb

require 'socket'
require 'surro-gate'

puts "proxying 1111 to 1112"
proxy = SurroGate.new
a = TCPSocket.new('localhost', 1111)
b = TCPSocket.new('localhost', 1112)
proxy.push a,b

begin
  loop do
    proxy.select(1)
    proxy.each_ready do |left, right|
        right.write_nonblock left.read_nonblock 1024
    end
  end
rescue => ex
  puts "done with connection: #{ex.message}"
  proxy.pop a
  proxy.pop b
end
```

**Pull request merged**